src/instruments/src/ATC/index.tsx

```diff
@@ -0,0 +1,132 @@
+
+    // Check if the transponder code simvar has changed. If it has, and we are not currently editing the code,
+    // set the display to the new code.
+    useEffect(() => {
+        if (currentEnteredDigits === 4) {
+            const newDigits = getDigitsFromCode(transponderCode);
+            if (newDigits.length === 4) {
+                setCurrentDigits(newDigits);
+            }
+        }
+    }, [transponderCode]);
+
+    // Count down the timers, if they are in use
+    useUpdate((deltaTime) => {
+        if (doubleClrTimer > 0) {
+            setDoubleClrTimer(doubleClrTimer - deltaTime / 1000);
+        }
+        if (displayResetTimer > 0) {
+            setDisplayResetTimer(displayResetTimer - deltaTime / 1000);
+        }
+    });
+
+    // Reset the double press timer, if the CLR button was not pressed a second time.
+    if (doubleClrTimer < 0 && doubleClrTimer !== -1) {
+        setDoubleClrTimer(-1);
+    }
+
+    // Check if less than 4 digits are currently entered, and start the reset timer.
+    if (displayResetTimer === -1 && currentEnteredDigits < 4) {
+        setDisplayResetTimer(5);
+    }
+    // Check if the reset timer has run up. If it has, reset the display to the previously entered transponder
+    // code, and reset the timer.
+    if (displayResetTimer < 0 && displayResetTimer !== -1) {
+        setDisplayResetTimer(-1);
+        const newDigits = getDigitsFromCode(transponderCode);
+        if (newDigits.length === 4) {
+            setCurrentDigits(newDigits);
+        }
+        setEnteredDigits(4);
+    }
+
+    const buttonPressHandler = (buttonId) => {
+        // Check if the pressed button is the reset button, and if ther is a digit displayed.
+        if (buttonId === 8 && currentEnteredDigits > 0) {
+            // If the clear button has not been pressed previously, just delete the rightmost digit.
+            if (doubleClrTimer === -1) {
+                setEnteredDigits(currentEnteredDigits - 1);
+                setDoubleClrTimer(0.2);
+                // If it has been pressed previously, clear the whole display.
+            } else {
+                setEnteredDigits(0);
+                setDoubleClrTimer(-1);
+            }
+            // If the pressed button is not the CLR button, and if not all digits have been entered yet,
+            // write to the next free digit.
+        } else if (buttonId !== 8 && currentEnteredDigits < 4) {
+            const digitsToSet = currentDigits;
+            digitsToSet[currentEnteredDigits] = buttonId;
+            setEnteredDigits(currentEnteredDigits + 1);
+            // If 3 digits are currently entered (i.e. if this digit will fill the display), write to the simvar.
+            if (currentEnteredDigits === 3) {
+                setTransponderCode(getBco16FromDigits(currentDigits));
+                setDisplayResetTimer(-1);
+            }
```

Comment:
Simplified and new variables:

```suggestion
            if(newDigits === null) {
                // currently still showing the normal transponder code - let's start with the first digit:
                setNewDigits([buttonId]);
            }
            else if (newDigits.length == 3) {
                // we are on our last digit - clear the draft and set the SimVar itself
                setNewDigits(null);
                setTransponderCode(getBco16FromDigits(newDigits)); // new digits is still available here until the next render
                setDisplayResetTimer(-1);
            }
            else {
                // we're on digit 0-3
                setNewDigits([...newDigits, buttonId]);
            }
```